engine/vcs/vcs.go

```diff
@@ -499,8 +499,11 @@ func RepoRootForImportDynamic(importPath string, verbose bool) (*RepoRoot, error
 		return nil, fmt.Errorf("%s: invalid repo root %q: %w", urlStr, metaImport.RepoRoot, err)
 	}
 	rr := &RepoRoot{
-		VCS:  ByCmd(metaImport.VCS),
-		Repo: metaImport.RepoRoot,
+		VCS: ByCmd(metaImport.VCS),
+		// ensure that dynamic discovery does not contain .git
+		// 99% of providers do not work with .git in URL
+		// e.g. Bitbucket, GitLab
+		Repo: strings.TrimSuffix(metaImport.RepoRoot, ".git"),
```

Comment:
Hm, we shouldn't strip this off right? It's part of the specified URL.

The user shouldn't specify `.git` on a GitLab URL if it's not supported.

Reply:
Mmh, I kind of agree but the current behavior of `RepoRootForImportPath` currently surprises me.

I feel more comfortable to have a coherent and expected behavior (detailed below), and be a bit smarter in this part of the logic instead of potentially paying the price later and handling those edge cases inside the `ModuleSource` part of the codebase.

A nice and easy mental model where:
- `Repo` represents a URL that a user can use to access the URL of the module, and all its variation (commit, tree view)
- `Root` represents the name of the module source. It is currently only being used as the name of the module. But being coherent on whether the user input ref contained a `.git` makes it a valid source of truth of the input ref (and would help us debug)

Given this heuristic:
- a repo should never contain a `.git`, whether it was specified or not
- a root should contain the `.git` if it was provided

#### Why is it important not to have the `.git` in the repo
Internally, `Repo` represents the URL to access the repo from a given ref. Upon manual tests, GitHub does support URLs with the `.git` inside ; but, most VCS do not, especially when trying to query the `tree` view of the commit (bitbucket / Gitlab)

#### Why is it important to keep track of the `.git`
If we follow the Go convention for private repo handling, and based on the current implementation of the `RepoRootForImportPath` it is recommended to use `.git` in the ref for private repos (due to GitLab returning fake metadata on `go-get=1` for unauthenticated refs). 

With this postulate, relying on this declarative model empowers the user and facilitates the handling on our side.

_What is the impact_ 
When having a ref with `.git`, the way to separate the root of the repo and potential subpath changes:
- without it, it uses HTTP requests
- with it, it uses the git CLI

I'll dig the implications a bit further on your https://github.com/dagger/dagger/pull/7708#issuecomment-2186297919. But, a nice way of handling root and subpaths is to ask users to actually specify the `.git`. This enables us to have ways to load auth at the `vcs.RepoRootForImportPath`

Reply:
> Left a couple nit comments, I'm not 100% sure about the .git stripping - if the VCS doesn't provide it, we should let the VCS error out with it, and not be super clever here.

Plus, I don't think it is just being clever: Google's logic for GitLab is currently a bit weird -> https://github.com/dagger/dagger/pull/7663/files#diff-a0ad18eec985477f121c56ba61b7956832670d35b67db059eac6f429085386b0L159. Here, the user referenced a ref without a `.git`, the metadata parsing from GitLab contained a `.git`, but as stated above, GitLab does not allow URLs with tree view on a commit when the URL contains a `.git`

Reply:
Honestly, fair enough point :tada: I'm happy with it as it! There's maybe *some* edge cases where this might not work, e.g. if `dagger.git` and `dagger` were separate repos - but that feels sooooo edge-casey, that we can fix it when we get to it.

Thanks for all the work on this, will be awesome to have tests for this!